[[ ScriptExtensions ]] Support script library extension loading

livecodeali opened the pull request:
This patch adds support for loading script library extensions in
the same way as lcb extensions are loaded. Metadata is retrieved
from the docs of the extension.

This patch removes the hard-coding of loading and docs generation
of current script library extensions in the IDE; such hard-coding
will no longer be necessary after companion PRs are merged.

montegoulding commented:
@livecodeali would it be a good idea to move these to components at the same time? At least in the commercial repo

livecodeali commented:
I'm not sure... are they components? Are components and extensions the same thing?

montegoulding commented:
pretty much... at least in the commercial repo anything that we might want to pick and choose which edition it should go in should probably be a component. Whether we need to organise things in such a way in the public repo is debatable though. Ideally we would end up with just one way to define a modular thing though.

livecodeali committed:
[[ ScriptExtensions ]] Don't load script library extensions manually

livecodeali committed:
[[ ScriptExtensions ]] Add support for loading script library extensions

livecodeali committed:
[[ ScriptExtensions ]] Build script extension docs dynamically

livecodeali committed:
[[ ScriptExtensions ]] Allow library extensions to use an svg icon

livecodeali committed:
[[ ScriptExtensions ]] Use com.livecode.library prefix for script extensions

livecodeali committed:
[[ ScriptExtensions ]] Tweak load order to support script extensions
Initialising a script extension with dependencies requires use of
revSBLibrary, which in turn requires a couple of handlers from
the backscript library. This patch moves these handlers to the
common library and moves it up the load order.

livecodeali committed:
[[ ScriptExtensions ]] Add handler for extension inclusion file data

livecodeali commented:
Depends on https://github.com/livecode/livecode/pull/5993

livecodeali committed:
[[ ScriptExtension ]] Load extensions from repo source
Load extensions from repo sources rather than the built
packaged_extensions folder. This prevents name clashes when using
the extension builder. The mechanism recursively searches the
given folders until it finds a candidate for an extension.

livecodeali committed:
[[ ScriptExtensions ]] Add support for loading script extensions using extension builder
Also sort and disambiguate duplicates in the extension builder
dropdown. This partially fixes bug 19946.

livecodeali committed:
[[ ScriptExtensions ]] Make sure extension manager displays data correctly
In particular, offer to uninstall user extensions and show the path
to the extension where the name is not available.

livecodeali committed:
[[ ScriptExtensions ]] Add release notes

montegoulding commented on notes/feature-script_extensions.md:
```diff
@@ -0,0 +1,88 @@
+
+## revLoadLibrary and revUnloadLibrary messages
+
+`revLoadLibrary` and `revUnloadLibrary` handlers should be implemented
```
@livecodeali I wonder if this is an opportune time to rename rev* stuff?

livecodeali replied:
Perhaps yeah... just `loadLibrary`?

montegoulding replied:
hmm.... well script extensions might feasibly be something other than libraries I guess. `loadExtension, unloadExtension` would be nice but then it might be expected to appear in `loadedExtensions`?

All having the same prefix might be nice. Perhaps:
- `extensionInitialize`
- `extensionTeardown`
- `extensionStartupScript`

montegoulding commented on Toolset/libraries/revidedeveloperextensionlibrary.livecodescript:
```diff
@@ -568,37 +575,46 @@ on revIDEDeveloperExtensionUninstall pPath
       if tEditor is not empty then
-         launch tFile with tEditor
+         launch pFile with tEditor
       end if
```
@livecodeali this seems a little odd that the OS pref would take precedence over the local IDE pref

livecodeali replied:
hmm, yeah, might be worth filing a bug report.

montegoulding commented on Toolset/libraries/revidedeveloperextensionlibrary.livecodescript:
```diff
@@ -568,37 +575,46 @@ on revIDEDeveloperExtensionUninstall pPath
       answer file "Select text editor..."
       if it is not empty then
-         launch tFile with it
+         launch pFile with it
```
Shouldn't this set the pref above?

livecodeali replied:
It does, below, if the launch was successful

montegoulding commented on Toolset/libraries/revideextensionlibrary.livecodescript:
```diff
@@ -838,35 +862,82 @@ function revIDEExtensionsOrderByDependency pExtensions
+   end if
+
+   put tExtFile into rSource
+   put tType into rType
```
@livecodeali if `tExtFile` is still empty here this would set `rSource` & `rType` incorrectly. Perhaps a type of `none`?

montegoulding commented on Toolset/libraries/revideextensionlibrary.livecodescript:
```diff
@@ -838,35 +862,82 @@ function revIDEExtensionsOrderByDependency pExtensions
+      put __fetchScriptLibraryData(tExtSource) into tArray
+   end if
+   
+   put tExtFile into tArray["source_file"]
```
I think `revIDEExtensionFetchSourceFromFolder` may feasibly set this to a list of files. Perhaps it should be an error if there is something other than a single file found in `revIDEExtensionFetchSourceFromFolder`?

livecodeali replied:
It does do that, above: https://github.com/livecode/livecode-ide/pull/1773/files/fca791decc95eb8c276ae94edbcf14b28baccc4c#diff-5a54288a32499468dd2cab9f178bed1dR891

livecodeali committed:
[[ DatagridDocs ]] Fix description parsing

(cherry picked from commit 474917cda43c8e931233abe4b1822da1d7feedda)

livecodeali committed:
[[ ScriptExtensions ]] Ensure type is set to 'none' if no ext file found

livecodeali committed:
[[ ScriptExtensions ]] Change revLoadLibrary and revUnloadLibrary
to extensionInitialize and extensionFinalize respectively. Also
change revLibraryStartupScript to extensionStartupScript

livecodeali committed:
[[ ScriptExtensions ]] Fix datagrid docs

livecodeali committed:
[[ ScriptExtensions ]] Fix load order
At the moment, all IDE libs that need to be loaded before any
extensions are before revIDEExtensionLibrary in the list.

montegoulding commented:
@livecode-vulcan review ok 7249bfe

Once merged we will need updates to mergJSON library and tsNet driver to use the new `extension*` API.

livecode-vulcan commented:
💙 review by @montegoulding ok 7249bfe86813535b0d8b2faab1938d6fa6a84ba0

livecodeali commented:
@montegoulding We don't currently fork those for develop do we? Probably the best bet is to keep the `rev` ones and add identical `extension` ones until we are no longer building 8.x

montegoulding commented:
Ah, yes... hmm would it be worth doing something like this for the time being?:
```
try
    send "extensionInitialize" to stack tStackName
catch tError
    send "revLoadLibary" to stack tStackName
end try
```

livecodeali commented:
Hmm, yeah that might be the best idea

livecodeali committed:
[[ ScriptExtensions ]] Try sending legacy messages
Try sending revLoadLibrary before extensionIntialize until we
drop support for LC 8.

montegoulding commented:
@livecode-vulcan review ok 4c5349e

livecode-vulcan commented:
💙 review by @montegoulding ok 4c5349ec415280335bcd6c1898e60e2f5a871262

livecodeali committed:
[[ ScriptExtensions ]] Change some extension stack names

livecodeali committed:
[[ ScriptExtensions ]] Use dispatch so we can detect if the
message was handled by the target or not.

montegoulding commented:
@livecode-vulcan review ok 9360372

livecode-vulcan commented:
💙 review by @montegoulding ok 9360372d72857b93b0b839a9e42d094363ca8df7

livecodeali committed:
Merge branch 'develop' into script-extensions

montegoulding commented:
@livecode-vulcan review ok 6d6bedd

livecode-vulcan commented:
💙 review by @montegoulding ok 6d6bedda8f80222f59e42016c20a071e62f006e7
